### PR TITLE
[CIT-468] SM Cumulocity Mapper

### PR DESCRIPTION
**Pull request by rina23q**
## Proposed changes

Implement Software Management Cumulocity Mapper.

How to run?
```
./tedge_mapper sm-c8y
```

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactorings that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by makr11st — COMMENTED** [review]
Added some comments, but I'll have another round after morning meetings.

I made some minor comments and added suggestion on quotes addition for reason string.
Also not required but a nice to have would be splitting serde portion of the code into separate crate similarly to `json_sm`.

I think we are missing service to be able to use it with `systemd` as well as deb metadata in Cargo.toml.

File: mapper/tedge_mapper/src/sm_c8y_mapper/smartrest_serializer.rs [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+    let mut s_clone = s.to_string();
+    let position_first_comma = s_clone
+        .find(',')
+        .ok_or(SmartRestSerializerError::DoubleQuoteError)?;
+    let s_after_comma = s_clone.split_off(position_first_comma + 1);
+    let position_second_comma = s_after_comma
+        .find(',')
+        .ok_or(SmartRestSerializerError::DoubleQuoteError)?;
+    Ok(position_first_comma + position_second_comma + 1)
+}
```
Comment (makr11st) [inline mapper/tedge_mapper/src/sm_c8y_mapper/smartrest_serializer.rs:222]: I think if you use `serde` as per above comment you may not need those functions anymore.

File: mapper/tedge_mapper/src/sm_c8y_mapper/error.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    DoubleQuoteError,
+
+    #[error(transparent)]
+    CsvError(#[from] csv::Error),
+
+    #[error(transparent)]
+    CsvWriterError(#[from] csv::IntoInnerError<Writer<Vec<u8>>>),
+
+    #[error(transparent)]
+    FromUtf8Error(#[from] std::string::FromUtf8Error),
```
[the remark this answers is not in the document]
Comment (rina23q) [reply mapper/tedge_mapper/src/sm_c8y_mapper/error.rs:25]: I agree to drop "Error", it's redundant although we use "XXXError" in lots of places of our code.
However, for example, `CSV` is also not good, should be `InvalidCSV` or something like that. This Errors looks very good examples https://doc.rust-lang.org/src/std/io/error.rs.html.

[the remark this answers is not in the document]
Comment (rina23q) [reply mapper/tedge_mapper/src/sm_c8y_mapper/error.rs:25]: But I'd like to keep some "Error" if the original error named with "Error", e.g. 'FromUtf8Error`.

File: mapper/tedge_mapper/src/sm_c8y_mapper/smartrest_serializer.rs [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+
+impl<'a> Serializer<'a> for SmartRestSetOperationToSuccessful {}
+
+#[derive(Debug, Deserialize, Serialize, PartialEq)]
+pub(crate) struct SmartRestSetOperationToFailed {
+    pub message_id: &'static str,
+    pub operation: &'static str,
+    pub reason: String,
+}
```
[the remark this answers is not in the document]
Comment (rina23q) [reply mapper/tedge_mapper/src/sm_c8y_mapper/smartrest_serializer.rs:165]: Really good! Thanks!!!!

**Review by makr11st — APPROVED** [review]
LGTM with remark that we have outstanding to address:
* integration tests - WIP
* add service files and deb package

File: mapper/tedge_mapper/src/sm_c8y_mapper/mapper.rs [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+                }
+                SoftwareCommand::Cumulocity => {
+                    debug!("c8y");
+                    let () = self
+                        .forward_software_request(message.payload_str()?)
+                        .await?;
+                }
+                _ => error!("Unknown topic {}", message.topic.name),
+            }
+        }
```
[the remark this answers is not in the document]
Comment (rina23q) [reply mapper/tedge_mapper/src/sm_c8y_mapper/mapper.rs]: Yes, it happened... I would like to highlight this issue, but address later as we discussed ;)

**Comment by rina23q** [comment]
> LGTM with remark that we have outstanding to address:
> 
> * integration tests - WIP
> * add service files and deb package

Those will be addressed in separated PRs.

**Review by abelikt — APPROVED** [review]
So far no concerns from my side. I haven't gone through the code but ran already some tests with it

File: mapper/tedge_mapper/src/sm_c8y_mapper/topic.rs [hunk 5]
```diff
@@ -0,0 +1,102 @@
+use json_sm::{SoftwareListRequest, SoftwareUpdateRequest};
+use mqtt_client::Topic;
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) enum SoftwareCommand {
```
Comment (albinsuresh) [inline mapper/tedge_mapper/src/sm_c8y_mapper/topic.rs:5]: The name is slightly misleading and it's hard to identify what this enum represents. I believe what you wanted to capture in this enum are all the topics from which the mapper would receive messages, right? In that case, naming is something like `SoftwareManagementIncomingTopic` would have made it clearer I feel. Just a suggestion.

File: mapper/tedge_mapper/src/sm_c8y_mapper/topic.rs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    SoftwareList,
+    SoftwareUpdate,
+    Cumulocity,
+    UnknownCommand,
```
Comment (albinsuresh) [inline mapper/tedge_mapper/src/sm_c8y_mapper/topic.rs:9]: Instead of an `UnknownCommand` enum variant here, I feel this enum should implement the `TryFrom` trait instead of the `From` trait which supports returning an error from the `try_from()` method which you'd want to represent an unsupported topic.

File: mapper/tedge_mapper/src/sm_c8y_mapper/smartrest_deserializer.rs [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug)]
+enum CumulocitySoftwareUpdateActions {
+    Install,
+    Delete,
+    UnknownAction,
+}
+
+impl From<String> for CumulocitySoftwareUpdateActions {
```
Comment (albinsuresh) [inline mapper/tedge_mapper/src/sm_c8y_mapper/smartrest_deserializer.rs:13]: Using `TryFrom` instead of `From` would let us get rid of that `UnknownAction` enum variant and express the failure cases the Rust way with an `Result:Err`

File: mapper/tedge_mapper/src/sm_c8y_mapper/smartrest_serializer.rs [hunk 8]
```diff
@@ -0,0 +1,541 @@
...
+    Self: Serialize,
+{
+    fn to_smartrest(&self) -> Result<SmartRest, SmartRestSerializerError> {
+        serialize_smartrest(self)
+    }
+}
+
+#[derive(Debug, Deserialize, Serialize, PartialEq)]
+pub(crate) struct SmartRestSetSupportedOperations {
+    pub message_id: &'static str,
```
Comment (albinsuresh) [inline mapper/tedge_mapper/src/sm_c8y_mapper/smartrest_serializer.rs:35]: Why is `message_id` an &str and not a `u16` or something numeric type? Is it because the CSV parser can't distinguish between numbers and strings while parsing a CSV string?

**Review by albinsuresh — CHANGES_REQUESTED** [review]

File: mapper/tedge_mapper/src/sm_c8y_mapper/error.rs [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    FromCsvWriter(#[from] csv::IntoInnerError<csv::Writer<Vec<u8>>>),
+
+    #[error(transparent)]
+    FromUtf8Error(#[from] std::string::FromUtf8Error),
+}
+
+#[derive(thiserror::Error, Debug)]
+pub(crate) enum SmartRestDeserializerError {
+    #[error("The SmartREST message ID is not Update Software Operation (528).")]
+    NotUpdateSoftwareOperation,
```
Comment (albinsuresh) [inline mapper/tedge_mapper/src/sm_c8y_mapper/error.rs:27]:
```suggestion
    UnsupportedOperation,
```

**Comment by rina23q** [comment]
Some tasks are left:

- integration tests - WIP
- add service files and deb package - #381 
- version name corner cases if :: contained in version names. Like 1.0.0::23::debian:: (high prio)
- refactoring - especially struct validation (low prio, as it failed somewhere)

File: mapper/tedge_mapper/src/sm_c8y_mapper/error.rs [hunk 10]
```diff
@@ -23,20 +17,27 @@ pub(crate) enum SmartRestSerializerError {
...
+#[derive(thiserror::Error, Debug)]
+pub(crate) enum MapperTopicError {
+    #[error("Topic {topic} is unknown.")]
+    UnknownTopic { topic: String },
+}
+
 #[derive(thiserror::Error, Debug)]
 pub(crate) enum SMCumulocityMapperError {
+    #[error(transparent)]
+    InvalidTopicError(#[from] MapperTopicError),
```
Comment (albinsuresh) [inline mapper/tedge_mapper/src/sm_c8y_mapper/error.rs:39]: Didn't really understand why this enum redirection was done. Why not define that `UnknownTopic` enum variant here directly?

**Review by albinsuresh — APPROVED** [review]

File: mapper/tedge_mapper/src/sm_c8y_mapper/error.rs (hunk 10 above)
Comment (rina23q) [reply mapper/tedge_mapper/src/sm_c8y_mapper/error.rs:39]: I wanted to group errors per file. Unknown topic error comes from the file `topic.rs`. and SMCumulocityMapperError comes from the file `mapper.rs`.
